### PR TITLE
Changes to facilitate CFA-0.6

### PR DESCRIPTION
**Pull request by davidhassell**
Needs https://github.com/NCAS-CMS/cfdm/pull/223 to be merged before review.

* Introduces "fragment array" classes and updates `NetCDFRead` to deal with aggregation variables; but all of the CFA-specific code is in  `NetCDFRead` is temporarily disabled. 
* Reorganises `cf/data/` to group array classes together.
* Refactors `cf/data/creation.py` to move some "to dask" functionality to `Array` subclasses.
* Made minor changes  to `test_dsg.py` and `test_gathering.py` to check that the new framework works.
* 
The purpose of this PR is to get some CFA placeholders in, but not to use or test any of the CFA code. However, the `cf/data/` rearrangement and `creation.py` changes are used and it would be good to merge these before any further work gets done on those areas.

### CODE REVIEW

**Comment by sadielbartholomew** [comment]
@davidhassell https://github.com/NCAS-CMS/cfdm/pull/223 was recently merged, so I going to start reviewing this, but there are some merge conflicts and GitHub isn't letting me resolve them myself (which I might have done if they were trivial and non-controversial, just so I could start reviewing immediately). If you can resolve those ASAP then I'll get the review done. Thanks.

**Comment by davidhassell** [comment]
Hi Sadie - after a slightly scary command line intervention, I think it's all OK now ...

**Comment by sadielbartholomew** [comment]
> I think it's all OK now ...

Thanks David. From a quick eyeball of the changeset here, and running of some tests with the updated branch, it looks like the conflict resolution did indeed go fine. I'll start a proper review now and aim to finish and submit it by the end of the day.

File: cf/data/array/fullarray.py [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+        return self._get_component("shape")
+
+    def get_fill_value(self):
+        """Return the data array fill value.
+
+        .. versionadded:: TODODASKVER
+
+        :Returns:
+
+            asdasda TODO
```
Comment (sadielbartholomew) [inline cf/data/array/fullarray.py:182]:
```suggestion
            TODODASKDOCS
```

File: cf/data/array/fullarray.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+            return "Uninitialised"
+
+        return f"Filled with {fill_value!r}"
+
+    def _set_units(self):
+        """TODO.
+
+        .. versionadded:: TODODASKVER
+
+        To cfdm.Array ?
```
Comment (sadielbartholomew) [inline cf/data/array/fullarray.py:150]:
```suggestion
        TODODASKDOCS
```

unless you can remember where you were going with that thought :smile:

File: cf/data/array/fullarray.py [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+            shape : tuple
+                The data array's dimension sizes.
+
+                size : int
+                    Number of elements indexin the data array.
+
+                fill_value : scalar, optional
+
+        #        masked_all: `bool`
```
Comment (sadielbartholomew) [inline cf/data/array/fullarray.py:40]:
```suggestion
```

File: cf/data/array/fullarray.py [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+
+        TODODASKDOCS
+
+            :Parameters:
+
+                dtype : numpy.dtype
+                    The numpy data type of the data array.
+
+            shape : tuple
+                The data array's dimension sizes.
```
Comment (sadielbartholomew) [inline cf/data/array/fullarray.py:33]: I see the `TODODASKDOCS` above but might as well fix the parameters already listed...

```suggestion
                shape : tuple
                    The data array's dimension sizes.
```

File: cf/data/array/cfanetcdfarray.py [hunk 5]
```diff
@@ -0,0 +1,693 @@
...
+
+
+class CFANetCDFArray(NetCDFArray):
+    """A CFA aggregated array stored in a netCDF file.
+
+    .. versionadded:: TODODASKVER
+
+    """
+
+    def __new__(cls, *args, **kwargs):
```
Comment (sadielbartholomew) [inline cf/data/array/cfanetcdfarray.py:20]: +1 for nice use of `__new__`

**Review by sadielbartholomew — APPROVED** [review]
I've raised some minor comments, mostly regarding docstrings, but overall this is all good. Please merge once you've considered all of the minor feedback, and the below.

As part of reviewing I generated some UML diagrams of the overall `data` class and module structure using `pyreverse` to check that everything looked sensible in the 'new way'. It all looks good to me, but thought I'd share here in case you wanted to review them too (they are large images so you'll need to e.g. right click to open in a new tab and then explore by moving about and/or zooming in):

### Classes
classes_DASK-CFA-DATA

### Modules
packages_DASK-CFA-DATA

File: cf/__init__.py [hunk 6]
```diff
@@ -199,7 +199,7 @@
     )
 
 # Check the version of dask
-_minimum_vn = "2022.6.0"
+_minimum_vn = "2022.1.0"
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/__init__.py:202]: Certainly not safe! I'll be fixing that ....

File: cf/read_write/netcdf/netcdfread.py [hunk 7]
```diff
@@ -401,3 +425,52 @@ def _array_from_variable(self, ncvar):
...
+
+            ncvar: `str`
+
+            unpacked_dtype: `False` or `numpy.dtype`, optional
+
+            coord_ncvar: `str`, optional
+
+        :Returns:
+
+            `CFANetCDFArray`, `dict`
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/read_write/netcdf/netcdfread.py:449]: Compromise? `(CFANetCDFArray, dict)`

File: cf/data/data.py [hunk 8]
```diff
@@ -443,34 +446,117 @@ def __init__(
...
+        #
+        #            array = to_dask(array, chunks, **kwargs)
+        #
+        #        elif chunks != _DEFAULT_CHUNKS:
+        #            # The data is already a dask array
+        #            raise ValueError(
+        #                "Can't define chunks for dask input arrays. Consider "
+        #                "rechunking the dask array before initialisation, or "
+        #                "rechunking the Data after initialisation."
+        #            )
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/data/data.py:559]: +1

File: cf/data/array/umarray.py [hunk 9]
```diff
@@ -0,0 +1,665 @@
...
+                ignored.
+
+                Number of elements in the uncompressed array.
+
+            ndim: `int`
+                Deprecated at version TODODASKVER. If set will be
+                ignored.
+
+                The number of uncompressed array dimensions.
+
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/data/array/umarray.py:83]: Thanks - https://github.com/NCAS-CMS/cf-python/pull/476/commits/a068d2bb7062616e506947c8617b44a5b7f7b611

File: cf/read_write/netcdf/netcdfread.py [hunk 10]
```diff
@@ -227,9 +207,72 @@ def _create_data(
...
+                "CFA-0.4 functionality is still available at version 3.13.1."
+            )
+
+            # TODOCFA: The 'return' remains when the exception is
+            #          removed at v4.0.0.
+            return True
+
+        cfa_04 = attributes.get("cf_role") == "cfa_variable"
+        if cfa_04:
+            # TODOCFA: Modify this message for v4.0.0.
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/read_write/netcdf/netcdfread.py:266]: I was clearly confused! They're all `TODOCFA`s, now.

**Comment by davidhassell** [comment]
Hi Sadie, thanks for the thorough review - I've addressed all of your points. Those diagrams are excellent! They've got me thinking about a few things (like should `NetCDFFragmentArray` be a subclass of `NetCDFArray`, partly so that it can pick up active storage capability (https://github.com/NCAS-CMS/cf-python/issues/501)?), but we'll revisit  all that when we do the final, full implementation of CFA.

**Pull request merged**